File: 2017/haskell/day02.hs
-- Advent of Code 2017, day 2

module Main where

import System.Environment
import Control.Applicative

import FunParser

main :: IO ()
main = do
  args <- getArgs
  puzzle (head args)

puzzle :: String -> IO ()
puzzle fileName = do
  putStrLn "Advent of Code 2017, day 2"
  input <- readFile fileName
  let xs = parseAll pInput input
  putStrLn ("Part 1: " ++ show (part1 xs))
  putStrLn ("Part 2: " ++ show (part2 xs))

-- Parsing the input

pInput :: Parser [[Int]]
pInput = pLines (some natural)

-- Part 1

part1 :: [[Int]] -> Int
part1 xs = sum [maximum(l)-minimum(l) | l <- xs]

-- Part 2

evenDiv :: [Int] -> (Int,Int)
evenDiv l = head [(x,y) | x<-l, y<-l, x/=y, x `rem` y == 0]

part2 :: [[Int]] -> Int
part2 xs = sum [x `div` y | (x,y) <- map evenDiv xs]

File: 2017/julia/day02.jl
# Advent of Code 2017 - day 2

f = open("input2_r")
input = readlines(f)
close(f)

function parse_nums(s)
  [parse(Int,n) for n in split(s)]
end

spreadsheet = [parse_nums(s) for s in input]

function diff(l)
  maximum(l)-minimum(l)
end

checksum = sum([diff(l) for l in spreadsheet])

println("part 1: ", checksum)

function divides(x,y)
  y !=x && rem(y,x)==0
end

function even_div(l)
  for x in l
    for y in l
      if divides(x,y) return div(y,x) end
    end
  end
end

div_sum = sum([even_div(l) for l in spreadsheet])

println("part 2: ", div_sum)

